aospy/test/test_automate.py

```diff
@@ -0,0 +1,192 @@
+import builtins
+import sys
+
+import pytest
+
+from aospy import Var, Proj
+from aospy.automate import (_get_attr_by_tag, _permuted_dicts_of_specs,
+                            _get_all_objs_of_type, _merge_dicts,
+                            _input_func_py2_py3, AospyException,
+                            _user_verify, CalcSuite, _MODELS_STR, _RUNS_STR)
+from .data.objects import examples as lib
+from .data.objects.examples import (
+    example_proj, example_model, example_run, condensation_rain,
+    convection_rain, precip, ps, sphum, globe, sahel
+)
+
+
+@pytest.fixture
+def obj_lib():
+    return lib
+
+
+@pytest.fixture
+def all_vars():
+    return {condensation_rain, convection_rain, precip, ps, sphum}
+
+
+@pytest.fixture
+def all_projects():
+    return {example_proj}
+
+
+@pytest.fixture
+def all_models():
+    return {example_model}
+
+
+@pytest.fixture
+def all_runs():
+    return {example_run}
+
+
+@pytest.fixture
+def all_regions():
+    return {globe, sahel}
+
+
+@pytest.fixture
+def calcsuite_specs():
+    return {
+        'time_offset': [None],
+        'date_range': ['default'],
+        'intvl_in': ['monthly'],
+        'region': [(globe, sahel)],
+        'dtype_out_time': [('av', 'reg.av')],
+        'dtype_in_vert': [False],
+        'dtype_in_time': ['ts'],
+        'var': [condensation_rain, convection_rain],
+        'intvl_out': ['ann'],
+        'dtype_out_vert': [None]
+    }
+
+
+@pytest.mark.parametrize(
+    ('obj', 'tag', 'attr_name', 'expected'),
+    [(example_proj, 'all', _MODELS_STR, [example_model]),
+     (example_proj, 'default', _MODELS_STR, [example_model]),
+     (example_model, 'all', _RUNS_STR, [example_run]),
+     (example_model, 'default', _RUNS_STR, [])])
```

Comment:
Yes, we should definitely be consistent.  I am leaning towards default returning empty if it hasn't been set.  That feels more intuitive.  But I could be convinced the other way, if you feel strongly about it.